WalletWasabi.Fluent/ViewModels/Dialogs/ShowQrCameraDialogViewModel.cs

```diff
@@ -31,51 +45,62 @@ protected override void OnNavigatedTo(bool isInHistory, CompositeDisposable disp
 	{
 		base.OnNavigatedTo(isInHistory, disposables);
 
-		Observable.FromEventPattern<Bitmap>(_qrReader, nameof(_qrReader.NewImageArrived))
-			.ObserveOn(RxApp.MainThreadScheduler)
-			.Subscribe(args => QrImage = args.EventArgs)
-			.DisposeWith(disposables);
-
-		Observable.FromEventPattern<string>(_qrReader, nameof(_qrReader.CorrectAddressFound))
-			.ObserveOn(RxApp.MainThreadScheduler)
-			.Subscribe(args => Close(DialogResultKind.Normal, args.EventArgs))
-			.DisposeWith(disposables);
-
-		Observable.FromEventPattern<string>(_qrReader, nameof(_qrReader.InvalidAddressFound))
-			.ObserveOn(RxApp.MainThreadScheduler)
-			.Subscribe(args =>
+		new CaptureDevices()
+			.EnumerateDescriptors()
+			.ToObservable()
+			.Skip(1)
+			.SelectMany(static d => d.Characteristics, static (d, c) => new { d, c })
+			.FirstOrDefaultAsync()
+			.Select(static d => d ?? throw new InvalidOperationException("Could not find a device."))
+			.Select(static d => d.d
+				.AsObservableAsync(d.c)
+				.ToObservable()
+				.Select(static d => d
+					.StartAsync()
+					.ToObservable()
+					.CombineLatest(d, static (_, scope) => scope))
+				.Switch())
+			.Switch()
+			.Select(scope =>
 			{
-				ErrorMessage = "No valid Bitcoin address found";
-				QrContent = args.EventArgs;
-			})
-			.DisposeWith(disposables);
+				var decoded = Decode(scope);
+				var bitmap = AddressStringParser.TryParse(decoded, _network, out _)
+					? null // No bitmap is required when the dialog is about to close
+					: new Bitmap(scope.Buffer.ReferImage().AsStream());
 
-		Observable.FromEventPattern<Exception>(_qrReader, nameof(_qrReader.ErrorOccurred))
+				return (decoded, bitmap);
+			})
```

Comment:
IMO this part should be encapsulated with the `Decode()` method in a Model class. Or an extension method.
So normally the ViewModel would just listen to that model class and set the properties once any data is received.

So basically everything that is related to reading could go to a separate class, And what is in the `Subscribe()` call here could stay here as that is actual logic that is related to the ShowQrCameraDialogVIewModel class.

Reply:
Made a model as it's for QR generation.